server/src/main/java/org/elasticsearch/search/vectors/KnnRescoreVectorQuery.java

```diff
@@ -0,0 +1,151 @@
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityByteValueSource;
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityFloatValueSource;
+import org.elasticsearch.search.profile.query.QueryProfiler;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Wraps a kNN vector query to rescore the results using the non-quantized vectors
+ */
+public class KnnRescoreVectorQuery extends Query implements ProfilingQuery {
+    private final String fieldName;
+    private final byte[] byteTarget;
+    private final float[] floatTarget;
+    private final VectorSimilarityFunction vectorSimilarityFunction;
+    private final Integer k;
+    private final Query vectorQuery;
+
+    private long vectorOpsCount;
+
+    public KnnRescoreVectorQuery(
+        String fieldName,
+        byte[] byteTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query vectorQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = byteTarget;
+        this.floatTarget = null;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.vectorQuery = vectorQuery;
+    }
+
+    public KnnRescoreVectorQuery(
+        String fieldName,
+        float[] floatTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query vectorQuery
+    ) {
+        this.fieldName = fieldName;
+        this.byteTarget = null;
+        this.floatTarget = floatTarget;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+        this.k = k;
+        this.vectorQuery = vectorQuery;
+    }
+
+    @Override
+    public Query rewrite(IndexSearcher searcher) throws IOException {
+        Query rewritten = super.rewrite(searcher);
+        if (rewritten != this) {
+            return rewritten;
+        }
+
+        final DoubleValuesSource valueSource;
+        if (byteTarget != null) {
+            valueSource = new VectorSimilarityByteValueSource(fieldName, byteTarget, vectorSimilarityFunction);
+        } else {
+            valueSource = new VectorSimilarityFloatValueSource(fieldName, floatTarget, vectorSimilarityFunction);
+        }
+        FunctionScoreQuery functionScoreQuery = new FunctionScoreQuery(vectorQuery, valueSource);
+        Query query = searcher.rewrite(functionScoreQuery);
+
+        if (k == null) {
+            // No need to calculate top k - let the request size limit the results
+            return query;
+        }
+
+        TopDocs topDocs = searcher.search(query, k);
+        ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+        int[] docIds = new int[scoreDocs.length];
+        float[] scores = new float[scoreDocs.length];
+        for (int i = 0; i < scoreDocs.length; i++) {
+            docIds[i] = scoreDocs[i].doc;
+            scores[i] = scoreDocs[i].score;
+        }
+
+        vectorOpsCount = scoreDocs.length;
```

Comment:
this seems like it should be the inner query vector ops + the total docs gathered.